tests/test_coordinate_source.py

```diff
@@ -0,0 +1,277 @@
+        # it is possible to override the particle size in the box file
+        src_new_size = EmanCoordinateSource(self.files_box, particle_size=100)
+        src_from_centers = CentersCoordinateSource(self.files_coord, particle_size=100)
+        imgs_new_size = src_new_size.images(0, 10)
+        imgs_from_centers = src_from_centers.images(0, 10)
+        for i in range(10):
+            self.assertTrue(np.array_equal(imgs_new_size[i], imgs_from_centers[i]))
+
+    def testImages(self):
+        # load from both the box format and the coord format
+        # ensure the images obtained are the same
+        src_from_coord = CentersCoordinateSource(self.files_coord, particle_size=256)
+        src_from_star = CentersCoordinateSource(self.files_star, particle_size=256)
+        src_from_relion = RelionCoordinateSource(
+            relion_autopick_star=os.path.join(
+                self.test_dir_root, "AutoPick/job006/sample_relion_autopick.star"
+            ),
+            particle_size=256,
+        )
+        imgs_box = self.src_from_box.images(0, 10)
+        imgs_coord = src_from_coord.images(0, 10)
+        imgs_star = src_from_star.images(0, 10)
+        imgs_relion = src_from_relion.images(0, 10)
+        for i in range(10):
+            self.assertTrue(np.array_equal(imgs_box[i], imgs_coord[i]))
+            self.assertTrue(np.array_equal(imgs_coord[i], imgs_star[i]))
+            self.assertTrue(np.array_equal(imgs_star[i], imgs_relion[i]))
+
+    def testImagesRandomIndices(self):
+        # ensure that we can load a specific, possibly out of order, list of
+        # indices, and that the result is in the order we asked for
+        # load images in known order
+        images_in_order = self.src_from_box.images(0, 440)
+        # shuffle all indices randomly
+        _indices = [i for i in range(440)]
+        random.shuffle(_indices)
+        images_random_order = self.src_from_box._images(indices=np.array(_indices))
+        for i, idx in enumerate(_indices):
+            self.assertTrue(
+                np.array_equal(images_in_order[idx], images_random_order[i])
+            )
+        # test loading every other image and compare
+        odd = np.array([i for i in range(1, 440, 2)])
+        even = np.array([i for i in range(0, 439, 2)])
+        odd_images = self.src_from_box._images(indices=odd)
+        even_images = self.src_from_box._images(indices=even)
+        for i in range(0, 220):
+            self.assertTrue(np.array_equal(images_in_order[2 * i], even_images[i]))
+            self.assertTrue(np.array_equal(images_in_order[2 * i + 1], odd_images[i]))
+
+        # random sample of [0,440) of length 100
+        random_sample = np.array(random.sample(_indices, 100))
+        random_images = self.src_from_box._images(indices=random_sample)
+        for i, idx in enumerate(random_sample):
+            self.assertTrue(np.array_equal(images_in_order[idx], random_images[i]))
+
+    def testMaxRows(self):
```

Comment:
We should probably also test the particle_size filtering in conjunction with the max_rows.

Reply:
Do you still want to test this in light of our decision to not factor in `max_rows` when excluding boundary particles? They happen in sequential order, boundary excluding followed by the `max_rows` cutoff

Reply:
There are a couple cases though right? defaults. max_rows < or > all available. max_rows < or > excluded avail. If the code works now, part of the reason to add the tests it to ensure if someone alters your code that they stay working, or at least don't crash :).

If you think its a lot of work we can skip it, at least unless it comes up again.

Reply:
I didn't even think of those. You're right, will add a few more. Thanks!